### PR TITLE
WIP: Fill in RenderTerraform methods for private topology

### PR DESCRIPTION
**Pull request by icereval**
Several `awstasks` are missing `RenderTerraform` methods, this is an attempt to fill them in so that we can render a cluster with private networking to Terraform output.

The work is kind of rough as we are not very familiar with the internals of the project, feel free to make changes or send feedback.

Might be a non-issue once #1055 is resolved.

NOTE: [loadbalancer_attributes.go](https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/awstasks/loadbalancer_attributes.go) still needs Terraform output when using the bastion. Terraform only supports inserting attributes in the `aws_elb` definition [same situation as health checks](https://github.com/kubernetes/kops/pull/1114/files#diff-00b0d5945b4937a8f4a4ef5bde97975aR193).

cc @jmcarp

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1114)

### CODE REVIEW

**Comment by kris-nova** [comment]
@k8s-bot ok to test

File: upup/models/cloudup/_aws/topologies/_topology_private/network.yaml [hunk 1]
```diff
@@ -190,6 +190,13 @@ loadBalancer/api.{{ ClusterName }}:
   {{ end }}
   listeners:
     443: { instancePort: 443 }
+  scheme: internal
```
Comment (justinsb) [inline upup/models/cloudup/_aws/topologies/_topology_private/network.yaml]: Changing the default scheme is a bigger change, and probably more controversial.  Can we split it out into another PR?

File: upup/pkg/fi/cloudup/awstasks/load_balancer.go [hunk 2]
```diff
@@ -323,3 +331,79 @@ func (_ *LoadBalancer) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *LoadBalan
 
 	return t.AddELBTags(*e.ID, t.Cloud.BuildTags(e.Name))
 }
+
+type terraformLoadBalancer struct {
+	Name           *string                           `json:"name"`
+	Listener       []*terraformLoadBalancerListener  `json:"listener"`
+	SecurityGroups []*terraform.Literal              `json:"security_groups"`
+	Subnets        []*terraform.Literal              `json:"subnets"`
+	Internal       bool                              `json:"internal,omitempty"`
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/load_balancer.go]: I think you want `*bool` if you want omitempty.

File: upup/pkg/fi/cloudup/awstasks/load_balancer.go [hunk 3]
```diff
@@ -323,3 +331,79 @@ func (_ *LoadBalancer) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *LoadBalan
...
+
+func (_ *LoadBalancer) RenderTerraform(t *terraform.TerraformTarget, a, e, changes *LoadBalancer) error {
+	elbName := e.ID
+	if elbName == nil {
+		elbName = e.Name
+	}
+
+	tf := &terraformLoadBalancer{
+		Name:     elbName,
+		Internal: *e.Scheme == "internal",
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/load_balancer.go]: I suspect `fi.StringValue(e.Schema) == "internal"` would be safer, in case Scheme is not set.  I presume the default value is a public ELB.

File: upup/pkg/fi/cloudup/awstasks/load_balancer.go [hunk 4]
```diff
@@ -323,3 +331,79 @@ func (_ *LoadBalancer) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *LoadBalan
...
+		Interval:           e.HealthCheck.Interval,
+		Timeout:            e.HealthCheck.Timeout,
+	}
+
+	return t.RenderResource("aws_elb", *e.Name, tf)
+}
+
+func (e *LoadBalancer) TerraformLink(params ...string) *terraform.Literal {
+	prop := "id"
+	if len(params) > 0 {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/load_balancer.go]: I don't love the varargs params, but I see what you're done and it is a clever way to do optional parameters in go.  Let's see where it goes - this isn't exposed externally.

File: upup/pkg/fi/cloudup/awstasks/load_balancer_attachment.go [hunk 5]
```diff
@@ -125,3 +126,33 @@ func (_ *LoadBalancerAttachment) RenderAWS(t *awsup.AWSAPITarget, a, e, changes
...
+func (_ *LoadBalancerAttachment) RenderTerraform(t *terraform.TerraformTarget, a, e, changes *LoadBalancerAttachment) error {
+	tf := &terraformLoadBalancerAttachment{
+		ELB: e.LoadBalancer.TerraformLink(),
+	}
+
+	if e.AutoscalingGroup != nil && e.Instance == nil {
+		tf.AutoscalingGroup = e.AutoscalingGroup.TerraformLink()
+		return t.RenderResource("aws_autoscaling_attachment", *e.AutoscalingGroup.Name, tf)
+	} else if e.AutoscalingGroup == nil && e.Instance != nil {
+		tf.Instance = e.Instance.TerraformLink()
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/load_balancer_attachment.go]: (We might want to deprecate instance attachment, but not in this PR.  I don't think anyone uses it...)

File: upup/pkg/fi/cloudup/awstasks/load_balancer_attachment.go [hunk 6]
```diff
@@ -125,3 +126,33 @@ func (_ *LoadBalancerAttachment) RenderAWS(t *awsup.AWSAPITarget, a, e, changes
...
+		return t.RenderResource("aws_elb_attachment", *e.LoadBalancer.Name, tf)
+	}
+	return nil
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/awstasks/load_balancer_attachment.go]: Probably better to return an error in this case - we don't expect this to happen I don't think

File: upup/models/cloudup/_aws/topologies/_topology_private/network.yaml (hunk 1 above)
Comment (kris-nova) [inline upup/models/cloudup/_aws/topologies/_topology_private/network.yaml]: Yes, can we please remove this and we can cover the feature in another PR? For now ELB's are our main entry point into the network, so we need them exposed. Also this PR shouldn't change functionality, just map it to Terraform is all

**Comment by kris-nova** [comment]
@justinsb I left a note on how I think we should **not** be setting the ELB to internal. If we can get that one liner looked at, that is enough for me to feel confident in merging.

The OP @icereval (Which, btw thank you for this PR - I was actually getting ready to start coding it when I noticed you already did) still has the **WIP** flag in the title, we should respect that until we hear back on what they want to do.

@icereval, 

Are you okay with addresses at minimum the one change that @justinsb pointed out [here](https://github.com/kubernetes/kops/pull/1114#discussion_r91882956)? And then pinging us so we can merge? I think we are trying to get this out fairly soon if that is okay with you.

Cheers

**Comment by kris-nova** [comment]
Merging in favor of #1136 for the code change
